### ISSUE TITLE
SABnzbd 4.2.2 when using 7za to unpack creates files with incorrect permissions (600 instead of 666)

### ISSUE THREAD

**Issue by JustAndreww** [SynoCommunity/spksrc#6009]
### Is this a new Bug?

- [X] I checked that the bug hasn't been reported before

### Package Name

SABnzbd

### Package Version

4.2.2

### Device Model

DS920+

### Device Architecture

x86_64

### Firmware Version

DSM 7.2.1-69057 Update 4

### What happened?

When unpacking 7Z downloads resulted file will have 600 permissions instead of 666. So I can't open it in Windows in mounted network drive. Doesn't happen with RAR downloads.

Permissions are set as follows for the downloading folder:
![image](https://github.com/SynoCommunity/spksrc/assets/4374611/35dc9657-2e71-4e37-87b4-1af589e11731)

### Reproduction steps

Download some NZB with 7z inside. Unpack will finish OK, resulted file cannot be opened in the mounted Windows drive folder due to lack of access rights.
No errors or whatsoever in logs, everything seems to be fine.

### Install Log

```log
https://pastebin.com/dXQHvhaD
```

**Comment by mreid-tt** [comment]
@Safihre, could this may be an upstream issue?

@JustAndreww, can you indicate if this unpack function had the correct permissions in previous versions? It would be helpful in understanding if there was a specific version where it did work correctly.

**Comment by JustAndreww** [comment]
I would love to, but I have only recently installed Sabnzbd and never used it before, so I can't really compare.

**Comment by Safihre** [comment]
Is is a bug in p7zip. Unfortunately the version used by SynoCommunity is 16.02, where the 16 stands for _2016_.
p7zip hasn't been updated since.
We could just try to compile the [official 7zip](https://www.7-zip.org/download.html) (without the "p"), since it has good Unix support now.
I'm just not that good in all the compile magic so I didn't dare to do so yet...

**Comment by JustAndreww** [comment]
Just used pre-compiled x86_64 linux binary from 7zip website (`7zzs`) - replaced local `7za` from `/var/packages/sabnzbd/target/bin/7za`, but still permissions are wrong after unpacking.

Maybe some additional arguments needs to be passed to `popen` from sabnzbd? 

Also I think this issue would be quite easy to fix with post-unpack-script that will simply do `chmod 666` to all the content in the unpacking directory, but I don't know how to write that kind of scripts :(

**Comment by Safihre** [comment]
SABnzbd can do that (Config > Folders > Advanced Settings > Permissions), however, this breaks Synology persmission system so that's why it's disabled by default.

If you can find which command line argument it is, I can always add it.

**Comment by JustAndreww** [comment]
Thank you for the hint!

I will stick to that solution at the moment. 755 permissions looks OK to my use-case and as a temporary workaround.

**Comment by th0ma7** [comment]
> Is is a bug in p7zip. Unfortunately the version used by SynoCommunity is 16.02, where the 16 stands for _2016_.
> p7zip hasn't been updated since.
> We could just try to compile the [official 7zip](https://www.7-zip.org/download.html) (without the "p"), since it has good Unix support now.
> I'm just not that good in all the compile magic so I didn't dare to do so yet...

I can have a look at that in the coming days if it helps.

**Comment by Safihre** [comment]
Unfortunately it seems the new one also doesn't have any special flags (like unrar) to not set permissions..

**Comment by BenjV** [comment]
As far as I know does 7zip not set any permission at all.
So the file will get the default permission of the process running Sabnzbd depenading on where the file is put.

If you need other permissions then you should make sure the location of the file is on a "shared folder" and the "system internal user" that sabnzbd is using must have read/write access there.

And the user from your Windows client should also have permissions there to be able to access them..

**Comment by Safihre** [comment]
@BenjV this is not true for 7zip.
Just use this test nzb. It's a zip, which will use 7zip to unpack. 
https://github.com/sabnzbd/sabnzbd.github.io/blob/master/tests/test_zip.nzb

**Comment by BenjV** [comment]
You can only test this on a Synology and with SabNzbd, any other test will be useless.

Within a zip there is no permission information so 7zip cannot set any meaninfull permissions.
So the permission the files get depend on the system setting where 7zip is running.

And on a Synology that will only result in a workable way if you put that files on a location where another user can access them as I explained.

**Comment by Safihre** [comment]
@BenjV seriously, did you test it on your Synology? Because it does not work. 
Show us a screenshot of the permissions that you get for a zip file downloaded and extracted with SABnzbd.

**Comment by mreid-tt** [comment]
@Safihre, I've done a few tests with the test nzb you provided. It normally downloads and extracts to `downloads/complete/test_zip`, with permissions like these:

```
$ ls -all
total 4380
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      24 Feb 12 08:22 .
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      20 Feb 12 08:22 ..
-rw-------  1 sc-sabnzbd synocommunity 4482652 Nov 18  2016 test_zip.bin
```

If I download only I get the folder with a file `8288838_ZIP_FILE.zip` in it. Taking a look at the details, it looks like a fairly simple archive:

```
$ 7za l -slt 8288838_ZIP_FILE.zip 

7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.utf8,Utf16=on,HugeFiles=on,64 bits,2 CPUs x64)

Scanning the drive for archives:
1 file, 3624126 bytes (3540 KiB)

Listing archive: 8288838_ZIP_FILE.zip

--
Path = 8288838_ZIP_FILE.zip
Type = zip
Physical Size = 3624126

----------
Path = testfile.bin
Folder = -
Size = 4482652
Packed Size = 3623968
Modified = 2016-11-18 11:00:48
Created = 2017-03-15 08:26:47
Accessed = 2017-03-15 08:26:47
Attributes = AC
Encrypted = -
Comment = 
CRC = DC714B31
Method = Deflate
Host OS = FAT
Version = 20
Volume Index = 0
```

A manual extraction looks like this:

```
$ sudo -u sc-sabnzbd 7za x 8288838_ZIP_FILE.zip

7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.utf8,Utf16=on,HugeFiles=on,64 bits,2 CPUs x64)

Scanning the drive for archives:
1 file, 3624126 bytes (3540 KiB)

Extracting archive: 8288838_ZIP_FILE.zip
--
Path = 8288838_ZIP_FILE.zip
Type = zip
Physical Size = 3624126

Everything is Ok

Size:       4482652
Compressed: 3624126
```
```
$ ls -all
total 7920
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      64 Feb 12 08:14 .
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      20 Feb 12 07:54 ..
-rw-rw-rw-  1 sc-sabnzbd synocommunity 3624126 Feb 12 07:54 8288838_ZIP_FILE.zip
-rwx------  1 sc-sabnzbd synocommunity 4482652 Nov 18  2016 testfile.bin
```

As you can see the permissions are defaulting to the running user only. I can however override this forcibly like this:

```
$ sudo -u sc-sabnzbd sh -c 'umask 111 && 7za x 8288838_ZIP_FILE.zip'

7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.utf8,Utf16=on,HugeFiles=on,64 bits,2 CPUs x64)

Scanning the drive for archives:
1 file, 3624126 bytes (3540 KiB)

Extracting archive: 8288838_ZIP_FILE.zip
--
Path = 8288838_ZIP_FILE.zip
Type = zip
Physical Size = 3624126

Everything is Ok

Size:       4482652
Compressed: 3624126
```
```
$ ls -all
total 7920
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      64 Feb 12 08:20 .
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      20 Feb 12 07:54 ..
-rw-rw-rw-  1 sc-sabnzbd synocommunity 3624126 Feb 12 07:54 8288838_ZIP_FILE.zip
-rw-rw-rw-  1 sc-sabnzbd synocommunity 4482652 Nov 18  2016 testfile.bin
```

It's not the best method but I couldn't see any permissions setting options for `7za` on first glance.

**Comment by th0ma7** [comment]
Permissions would then follow the umask directive from the user shell environment...

**Comment by mreid-tt** [comment]
In my view, unless `7za` supports setting permissions via arguments then any sort of workaround (umask or chmod) would be better served by the existing mechanism for setting permissions post-download. It doesn't add value in my view to make yet another workaround.

**Comment by Safihre** [comment]
I thought so indeed.. Too bad 7zip doesn't have a parameter for it. Unrar does have a specific parameter to not set permissions (which we use).

**Comment by mreid-tt** [comment]
@Safihre, based on my understanding, it seems that this issue is not exclusive to Synology but rather could potentially affect any Linux implementation. Would you concur with this assessment? If so, it implies that the problem lies upstream and is not inherently a flaw within SABnzbd itself. This is due to a dependency constraint preventing a straightforward resolution, though there exists a workaround to mitigate the issue. Would you agree with this interpretation?

**Comment by Safihre** [comment]
Agreed. Not Synology specific. The complicating factor here is that if Permissions is set in SABnzbd Config Folders, it uses chmod and breaks the Synology permissions. But that's all we can do about it.

**Comment by th0ma7** [comment]
> So how would we choose the umask?

`umask` is rather interesting... consider the following on my linux, default behavior:
```
$ touch test
$ ls -la test
-rw-rw-r-- 1 th0ma7 th0ma7 0 fév 12 21:39 test
```
Mode is `u+rw, g+rw, a+r` which would be described as `0664` with chmod command using octal mode.  Likewise setting it `rwx` to all would be:
```
$ chmod 0777 test
$ ls -al test
-rwxrwxrwx 1 th0ma7 th0ma7 0 fév 12 21:39 test
```

Now what about umask you ask... well umask runs in octal mode where you mask certain bits.  For instance, a normal `u+rw, g+rw, a+r` ,described as 0664, would mean having a umask of 0113 such as:
```
  0777
- 0664
--------
  0113
```
So using this principle:
```
$ umask 0113   <-- equivalent of chmod 0664, or considering 0777-0664 = 0113
$ rm test
$ touch test
$ ls -la test
-rw-rw-r-- 1 th0ma7 th0ma7 0 fév 12 21:42 test
```
`umask` sets the file creation mode of, well, files to be created by said program for which its environment was set with a certain umask.  Now wanting files to be `rw` to all would be:
```
$ umask 0111   <-- equivalent of chmod 0666, or considering 0777-0666 = 0111
$ rm test
$ touch test
$ ls -la test
-rw-rw-rw- 1 th0ma7 th0ma7 0 fév 12 21:42 test
```
and so on... 

Now, going back to `sabnzbd`, adding a simple `umask 0113` (for 0664) or `umask 0111` (for 0666) in the startup script may resolve the issue.  Again, this may have a different behavior when used in conjunction with ACL ... but worth testing.

**Comment by BenjV** [comment]
You are all forgetting that Synology has implemented its own permission system based on the ACL's and not on the default Linux permissions.

**Comment by Safihre** [comment]
We are very well aware. 
So what do you propose we do @BenjV?

**Comment by BenjV** [comment]
I already explained it.

The target where 7zip will place the files must be on a "shared folder" and that share folder must be granted read/write privs for the "system internal user" which is running 7zip so in the case 7zip is spawned by SabNzbd this system intenal user of SabNzbd.

And if you need access to those files by a user from the network, that user must be granted privs also.
And all this granting must be done via the DSM gui otherwise it will not function correctly.

**Comment by mreid-tt** [comment]
@BenjV, isn't that the same advice we already provide in the [Permission Management](https://github.com/SynoCommunity/spksrc/wiki/Permission-Management) wiki? If so, then there is nothing to address in the package or the upstream software. If not, then I am not clear on how that addresses the issue raised by the user.

**Comment by Safihre** [comment]
@BenjV Do you really not read _anything_ posted by other users here? 
@mreid-tt [showed](https://github.com/SynoCommunity/spksrc/issues/6009#issuecomment-1938589046) that 7zip forcedly sets permissions, so it's an upstream problem of 7zip that always sets permissions and breaks the usual DSM permission system.

**Comment by th0ma7** [comment]
And if i may add, i played with 7zip to replace p7zip and wow ... Cross-compiling this isn't easy.  I wasn't able to get far, even following arch linux recipe.

**Comment by th0ma7** [comment]
Yes.  But testing is needed considering both acl and how p7zip works.

**Comment by Safihre** [comment]
Considering this is the first report of this problem even though the problem has existed for years, should we really be adding that?

**Comment by mreid-tt** [comment]
Sorry, I wasn't advocating a proposed change, I stand by my original recommendation that this is well handled by the existing post-processing. I was just clarifying in case I wanted to do some private testing (no commitment to so do).

**Comment by BenjV** [comment]
> @BenjV Do you really not read _anything_ posted by other users here? @mreid-tt [showed](https://github.com/SynoCommunity/spksrc/issues/6009#issuecomment-1938589046) that 7zip forcedly sets permissions, so it's an upstream problem of 7zip that always sets permissions and breaks the usual DSM permission system.

I could say the same of you.
The point is not to test this kind of thing on the command line because DSM is using ACL's and not the standard linux permission system.
So the permissions DSM is acting on, is depending on where the files are put by which user.
That is also the reason no package should set permissions by itself, that would mess-up things.

And as @mreid-tt said, this is already explained in the wiki.

**Comment by Safihre** [comment]
> If 7zip actualy sets the permission itself and by that messes-up the DSM permissions it can only be fixed upstreams, but I very much doubt that this is the case. A normal application does not do anything with permissions but will leave that to the permission of the user who is running the application.

What more proof do you need that 7zip sets permissions? 
We see it happen on the command line and when the process is spawned by SABnzbd.

**Comment by BenjV** [comment]
Did you read the results of the test @mreid-tt did?????
From his post:

> As you can see the permissions are defaulting to the running user only.

This is proof that 7zip does **not** set the permissions on a zip file just like a rar file.

What maybe is happening is that the zip file contatins ntfs permissions and 7zip support this.
If that command switch is used (-sni), then that could explain the different outcome between rar files and zip files.

**Comment by mreid-tt** [comment]
Folks, let's dial it back a bit. Both of you raise valid points, but I propose that instead of further speculation, we invest our time in conducting actual tests to validate our hypotheses. Here are a few points to consider:

@Safihre, as far as I can tell, 7zip does not set permissions for the archive in question. If this were the case the `umask` would not work as 7zip would override this environment setting.

@BenjV, the zip file in question does not contain any relevant NTFS permissions given the attributes that are set in the file were (as stated above):
```
Attributes = AC
```
Based on my observations, the attributes "archived and compressed" do not have any bearing on user permissions. Additionally, I tested using the `-sni` switch, but it did not alter the behaviour of the extracted file.

I am skeptical that permissions are solely inherited from the user and location. For instance, the parent folder of the zip file, created by 7zip, does not share the same permissions as its containing folder, as illustrated previously. My suspicion is that inheritance might be overridden, resulting in the absence of ACL+ attributes.

Before delving deeper into this discussion, I urge you to replicate the test I conducted and examine the folder permissions along with the ACL+ attributes. Further discourse without concrete evidence would be speculative.

**Comment by BenjV** [comment]
It probably only happens when the zip file contains ntfs permissions so  the zip has to be created on a Windows system also with the -sni argument.
This is the only explaination I can think of regarding the difference between a rar file and a zip file.

When I find time (tomorrow maybe) I will perform some test.

**Comment by BenjV** [comment]
Tested with a file containing ntfs permissions but on the Nas 7zip could not extract such a file, so that is not the problem.

**Comment by mreid-tt** [comment]
hey @Safihre, a bit off topic but I was testing a recent download of an NZB file which contained a RAR archive and I was surprised to find the extracted files with the following permissions:

```
drwxrwxrwx+ 1 sc-sabnzbd synocommunity     1572 Apr 17 09:04  .
drwxrwxrwx+ 1 sc-sabnzbd synocommunity       94 Apr 17 09:04  ..
-rw-------  1 sc-sabnzbd synocommunity      697 Apr 15 12:02  00-sam_smith-love_goes-web-2020.m3u
-rw-------  1 sc-sabnzbd synocommunity     8962 Apr 15 12:02  00-sam_smith-love_goes-web-2020.nfo
-rw-------  1 sc-sabnzbd synocommunity  6126566 Apr 15 12:02  01-sam_smith-young.mp3
-rw-------  1 sc-sabnzbd synocommunity  8560139 Apr 15 12:02  02-sam_smith-diamonds.mp3
[---snip---]
```

There were no ACLs set by default as this command shows:

```
synoacltool -get 02-sam_smith-diamonds.mp3 
(synoacltool.c, 596)It's Linux mode
```

Digging further I realised that while it was a RAR archive, on extraction it had another layer which was a ZIP file as shown below:

```
Archive: RT91vsW5g1zzV2yu0dUcPzAJAw1Hy.part01.rar
Details: RAR 5, volume 1

 Attributes      Size     Date    Time   Name
----------- ---------  ---------- -----  ----
 -rw-rw-rw- 135780369  2024-04-15 12:02  0dRErHHXynYen6GVBaM.zip
----------- ---------  ---------- -----  ----
            135780369                    1
```

As such, when the files were extracted with `7za` the same lack of permissions setting took place. Given that the files don't actually have any ACLs set I don't really believe that setting the permissions on completed downloads is an actual issue. Setting it to '777', the newly extracted directory contains:

```
drwxrwxrwx  1 sc-sabnzbd synocommunity     1572 Apr 17 09:33  .
drwxrwxrwx+ 1 sc-sabnzbd synocommunity      168 Apr 17 09:33  ..
-rw-rw-rw-  1 sc-sabnzbd synocommunity      697 Apr 15 12:02  00-sam_smith-love_goes-web-2020.m3u
-rw-rw-rw-  1 sc-sabnzbd synocommunity     8962 Apr 15 12:02  00-sam_smith-love_goes-web-2020.nfo
-rw-rw-rw-  1 sc-sabnzbd synocommunity  6126566 Apr 15 12:02  01-sam_smith-young.mp3
-rw-rw-rw-  1 sc-sabnzbd synocommunity  8560139 Apr 15 12:02  02-sam_smith-diamonds.mp3
[---snip---]
```

The ACL details are the same:

```
synoacltool -get 02-sam_smith-diamonds.mp3 
(synoacltool.c, 596)It's Linux mode
```

I believe this is a reasonable workaround which doesn't seem to cause any issues on Synology as far as I can see.
